---
title: "Assignment Week 8"
author: "Ruqia Ali Hassan"
date: "2025-04-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# ATAC-seq - differential motif accessibility

## Installing packages and loeading libraries
```{r, warning=FALSE}
suppressPackageStartupMessages({
  library(epiwraps)
  library(motifmatchr)
  library(MotifDb)
  library(ggplot2)
  library(universalmotif)
  library(SummarizedExperiment) # data structure
  library(sechm) # for plotting heatmaps from a SummrizedExperiment
  library(BiocParallel) # for multithreading
  library(chromVAR) # for motif accessibility estimation
  library(limma) # for statistical analysis
  library(stats)
  library(Rsamtools)
  library(BSgenome)
  library(rtracklayer)
  library(GenomicRanges)
  library(BSgenome.Mmusculus.UCSC.mm10)

})

register(SnowParam(2))
```

## Download the data
```{r}
download.file("https://ethz-ins.org/content/mouse_mm38_hippocampus.peakCounts.SE.rds", "ATAC-seq_peak_counts_hippocampus.rds")
se<-readRDS("ATAC-seq_peak_counts_hippocampus.rds")
```

## Getting the motif and genom sequence
```{r}
#getting the motifs
motifs<-query(MotifDb, c("Mmusculus", "HOCOMOCOv10"))

motifs <- do.call(TFBSTools::PWMatrixList, setNames(
           universalmotif::convert_motifs(motifs, class="TFBSTools-PWMatrix"),
           mcols(motifs)$geneSymbol))
#motifs

#getting the genom sequence
genome <- BSgenome.Mmusculus.UCSC.mm10
```

## Preparing the data
```{r}
colData(se)
sample_names <- rownames(colData(se))
colData(se)$condition <- ifelse(grepl("FSS", sample_names), "FSS", "Control")
colData(se)$sex <- ifelse(grepl("female", sample_names), "Female", "Male")

seqlevels(genome)
seqlevels(se)
se <- keepStandardChromosomes(se, pruning.mode="coarse")
seqlevelsStyle(se) <- "UCSC"
seqlevels(se)
```

### Adding GC bias 
```{r}
se <- chromVAR::addGCBias(se, genome=genome)
rowData(se)
hist(rowData(se)$bias)
```

## Peak analysis
```{r, warning=FALSE}
moi <- motifmatchr::matchMotifs(motifs, subject=se, genome=genome)
head(assay(moi))

# ensure reproducibility
set.seed(1234)
bg <- chromVAR::getBackgroundPeaks(se, niterations=500)
dev <- chromVAR::computeDeviations(object = se, annotations=moi,
                                   background_peaks=bg)
```

## Diffrential analysis

```{r}
metadata(dev)$anno_colors <- list(condition = c(Control = "lightgrey", FSS = "darkred"),
                                  sex = c(Female = "pink", Male = "lightblue"))

```

### Control vs. FSS
```{r}
dev$condition <- factor(dev$condition) 
mm_cond <- model.matrix(~condition, data=as.data.frame(colData(dev)))
```

```{r}
fit_cond <- limma::eBayes(limma::lmFit(assay(dev, "z"), design = mm_cond))
res_cond <- limma::topTable(fit_cond, coef = "conditionFSS", number = Inf)

unique_names <- make.unique(rownames(assay(dev, "z")))
rownames(res_cond) <- unique_names
head(res_cond)


sechm::sechm(dev, features = head(rownames(res_cond)), assayName = "z", top_annotation = c("condition", "sex"))
```

The ANDR motif, which is associated with androgen receptor signaling, is more accessible in the FSS group, suggesting that stress may influence pathways regulated by androgens. This is supported by a statistically significant adjusted p-value of 0.0016. While other motifs also exhibit differences in accessibility, these changes do not reach statistical significance, possibly reflecting more subtle alterations in chromatin structure within pathways that are less affected by stress under the conditions of this experiment.



### Female vs. Male 
```{r}
dev$sex
dev$sex <- factor(dev$sex)
mm_sex <- model.matrix(~sex, data = as.data.frame(colData(dev)))
```

```{r}

fit_sex <- limma::eBayes(limma::lmFit(assay(dev, "z"), mm_sex))
res_sex <- limma::topTable(fit_sex, coef = "sexMale", number = Inf)

unique_names <- make.unique(rownames(assay(dev, "z")))
rownames(res_sex) <- unique_names
head(res_sex)

sechm::sechm(dev, features = head(rownames(res_sex)), assayName = "z", top_annotation = c("condition", "sex"))
```

The ANDR motif also shows the most significant difference in accessibility between females and males (adjusted p-value = 0.000029). While other motifs display sex-based differences in accessibility, their adjusted p-values are higher, indicating lower statistical significance compared to ANDR. These results suggest the presence of sex-specific differences in the binding of key transcription factors, particularly those involved in hormone signaling.

Together, these findings highlight the impact of both stress and biological sex on chromatin accessibility, particularly within hormone signaling pathways. 

```{r cars}
summary(cars)
```

